File: edukate-common/src/main/java/io/github/sanyavertolet/edukate/common/notifications/CheckedNotificationCreateRequest.java
package io.github.sanyavertolet.edukate.common.notifications;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonTypeName;
import io.github.sanyavertolet.edukate.common.checks.CheckStatus;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import lombok.experimental.SuperBuilder;

@Getter
@Setter
@SuperBuilder
@ToString(callSuper = true)
@JsonTypeName("checked")
public final class CheckedNotificationCreateRequest extends BaseNotificationCreateRequest {
    private String submissionId;
    private String problemId;
    private CheckStatus status;

    @JsonCreator
    public CheckedNotificationCreateRequest(
            String uuid, String targetUserId, String submissionId, String problemId, CheckStatus status
    ) {
        super(uuid, targetUserId);
        this.submissionId = submissionId;
        this.problemId = problemId;
        this.status = status;
    }
}

File: edukate-common/src/main/kotlin/io/github/sanyavertolet/edukate/common/notifications/CheckedNotificationCreateRequest.kt
package io.github.sanyavertolet.edukate.common.notifications

import com.fasterxml.jackson.annotation.JsonTypeName
import io.github.sanyavertolet.edukate.common.checks.CheckStatus
import java.util.UUID

@JsonTypeName("checked")
data class CheckedNotificationCreateRequest(
    override val uuid: String = UUID.randomUUID().toString(),
    override val targetUserId: Long,
    val submissionId: Long,
    val problemKey: String,
    val status: CheckStatus,
) : BaseNotificationCreateRequest {
    companion object {
        @JvmStatic
        fun from(targetUserId: Long, submissionId: Long, problemKey: String, status: CheckStatus) =
            CheckedNotificationCreateRequest(
                targetUserId = targetUserId,
                submissionId = submissionId,
                problemKey = problemKey,
                status = status,
            )
    }
}
